### PR TITLE
Implement Agent Skills [1/N]: Create training skill (MVP)

### PR DESCRIPTION
**Pull request by albertvillanova**
Implement Agent Skills [1/N]: Create training skill (MVP)

Create main skill file containing comprehensive TRL CLI training instructions.

Part of:
- #5095

### Changes
- Create trl-training skill
- Update MANIFEST.in: include all markdown files under the `trl/skills/` directory and its subdirectories in the package distribution

### Future extensions
- Keep SKILL.md body under 5000 tokens (~20,000 chars)
- Give detailed examples in references/cli-examples.md
- Give complete YAML configuration examples for each training method in references/config-examples.yaml
- Give common errors and solutions in references/troubleshooting.md

### CODE REVIEW

File: trl/skills/trl-training/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
+---
+name: trl-training
+description: Train and fine-tune transformer language models using TRL (Transformer Reinforcement Learning). Supports SFT, DPO, GRPO, KTO, RLOO and Reward Model training via CLI commands.
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
description: Train and fine-tune transformer language models using TRL (Transformers Reinforcement Learning). Supports SFT, DPO, GRPO, KTO, RLOO and Reward Model training via CLI commands.
```

File: trl/skills/trl-training/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    - dpo
+    - grpo
+    - huggingface
+    - transformers
+  documentation: https://huggingface.co/docs/trl/en/clis
+---
+
+# TRL Training Skill
+
+You are an expert at using the TRL (Transformer Reinforcement Learning) library to train and fine-tune large language models.
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
You are an expert at using the TRL (Transformers Reinforcement Learning) library to train and fine-tune large language models.
```

File: trl/skills/trl-training/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+## Overview
+
+TRL provides CLI commands for post-training foundation models using state-of-the-art techniques:
+
+- **SFT** (Supervised Fine-Tuning): Fine-tune models on instruction-following or conversational datasets
+- **DPO** (Direct Preference Optimization): Align models using preference data without a reward model
+- **GRPO** (Group Relative Policy Optimization): Train models using LLM-as-a-judge for feedback
+- **KTO** (Kahneman-Tversky Optimization): Align models using binary feedback (good/bad)
+- **RLOO** (Reinforcement Learning with Language Model Objectives): Online RL training with generation-based rewards
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- **RLOO** (Reinforce Leave One Out): Online RL training with generation-based rewards
```

File: trl/skills/trl-training/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+You are an expert at using the TRL (Transformer Reinforcement Learning) library to train and fine-tune large language models.
+
+## Overview
+
+TRL provides CLI commands for post-training foundation models using state-of-the-art techniques:
+
+- **SFT** (Supervised Fine-Tuning): Fine-tune models on instruction-following or conversational datasets
+- **DPO** (Direct Preference Optimization): Align models using preference data without a reward model
+- **GRPO** (Group Relative Policy Optimization): Train models using LLM-as-a-judge for feedback
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- **GRPO** (Group Relative Policy Optimization):Train models by ranking multiple sampled outputs relative to each other and optimizing based on their comparative rewards.
```

File: trl/skills/trl-training/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+**Key parameters:**
+
+- `--model_name_or_path`: Model to train
+- `--dataset_name`: Training dataset with prompts
+- `--judge_model`: Judge model for scoring (e.g., "gpt-4", "claude-3-opus")
+- `--num_generations`: Number of generations per prompt (default: 4)
+- `--learning_rate`: Learning rate (default: 1.0e-6)
+
+**Dataset format:** Requires "prompt" column. Generations are scored by the judge model.
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
**Dataset format:** Requires "prompt" column. Generations are scored by the reward function
```

File: trl/skills/trl-training/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+  --dataset_name trl-lib/gsm8k \
+  --judge_model gpt-4 \
+  --output_dir ./grpo_output
+```
+
+**Key parameters:**
+
+- `--model_name_or_path`: Model to train
+- `--dataset_name`: Training dataset with prompts
+- `--judge_model`: Judge model for scoring (e.g., "gpt-4", "claude-3-opus")
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- `--reward_funcs`: Reward functions, one or more of "accuracy_reward", "reasoning_accuracy_reward", "think_format_reward", "get_soft_overlong_punishment"
```

File: trl/skills/trl-training/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+- `--num_generations`: Number of generations per prompt (default: 4)
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- `--num_generations`: Number of generations per prompt (default: 8)
```

File: trl/skills/trl-training/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+```json
+{"messages": [
+  {"role": "user", "content": "Hello!"},
+  {"role": "assistant", "content": "Hi! How can I help?"}
+]}
+```
+
+**Completion format:**
+```json
+{"prompt": "Translate to French: Hello", "completion": "Bonjour"}
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
{"prompt": "The translation to French of Hello is", "completion": " Bonjour."}
```

File: trl/skills/trl-training/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+  "chosen": "Silicon dreams...",
+  "rejected": "Beep boop robot..."
+}
+```
+
+### KTO Datasets
+
+**Binary feedback format:**
+```json
+{"prompt": "Explain gravity", "completion": "Gravity is...", "label": true}
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
{"prompt": [{"role": "user", "content": "Explain gravity."}], "completion": [{"role": "assistant", "content": "Gravity is..."}], "label": true}
```

File: trl/skills/trl-training/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+```
+
+### GRPO/RLOO Datasets
+
+**Prompt-only format:**
+```json
+{"prompt": "Solve: 2+2=?"}
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
{"prompt": [{"role": "user", "content": "Solve: 2+2=?"}], "solution": "4"}
```

File: trl/skills/trl-training/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+
+### 2. LoRA Fine-Tuning (Memory Efficient)
+
+```bash
+# Train with LoRA adapters
+trl sft \
+  --model_name_or_path meta-llama/Llama-3.2-1B \
+  --dataset_name trl-lib/Capybara \
+  --output_dir ./llama-lora \
+  --use_peft true \
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
  --use_peft \
```

File: trl/skills/trl-training/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+```
+
+### 4. GRPO with Custom Judge
+
+```bash
+# Train with LLM judge feedback
+trl grpo \
+  --model_name_or_path ./llama-sft \
+  --dataset_name trl-lib/gsm8k \
+  --judge_model gpt-4o \
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
  --reward_funcs accuracy_reward \
```

File: trl/skills/trl-training/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+  --num_train_epochs 1
+```
+
+### 4. GRPO with Custom Judge
+
+```bash
+# Train with LLM judge feedback
+trl grpo \
+  --model_name_or_path ./llama-sft \
+  --dataset_name trl-lib/gsm8k \
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
  --dataset_name trl-lib/DeepMath-103K \
```

File: trl/skills/trl-training/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+  --num_generations 8 \
+  --learning_rate 1.0e-6
+```
+
+## Troubleshooting
+
+### CUDA Out of Memory
+
+- Reduce `--per_device_train_batch_size`
+- Increase `--gradient_accumulation_steps`
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- Reduce `--per_device_train_batch_size` and increase `--gradient_accumulation_steps`
```

File: trl/skills/trl-training/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+- Enable `--tf32` for faster computation on Ampere GPUs
+- Use `--bf16` on supported hardware
+- Consider multi-GPU training with `--num_processes`
+
+### Generation Issues (GRPO/RLOO)
+
+- Check prompt format in dataset
+- Adjust `--temperature` and `--top_p` for generation
+- Verify judge model has API access (for GRPO)
+- Check reward model compatibility (for RLOO)
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]:
```suggestion
- Verify the reward function (for GRPO/RLOO)
```

**Comment by albertvillanova** [comment]
Thanks for your detailed review!

**Comment by albertvillanova** [comment]
@qgallouedec, thanks for catching these, and apologies for the oversights.

My goal with this PR was mainly to introduce a preliminary `SKILLS.md` so I could exercise and validate the related CLI functionality (installation, listing, etc.) that I'm building in the subsequent PRs. That said, I should have paid closer attention to the actual content of the file and avoided these obvious mistakes, even if it was meant as an initial scaffold.

I have addressed the suggested fixes. Thanks again for the careful review.

**Comment by albertvillanova** [comment]
I am reviewing this so this PR implements just minimal content.

File: trl/skills/trl-training/SKILL.md [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+- `--dataset_name`: Preference dataset
+- `--learning_rate`: Learning rate (default: 5.0e-7)
+- `--beta`: DPO temperature parameter (default: 0.1)
+- `--loss_type`: Loss function (default: "sigmoid"; options: "sigmoid", "hinge", "ipo", "bco_pair")
+
+**Dataset format:** Requires "prompt", "chosen", and "rejected" columns.
+
+### trl grpo - Group Relative Policy Optimization
+
+Train models using LLM-as-a-judge for evaluating generations and providing rewards.
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]: not 100% accurate as we usually use reward function in general, but we can improve later

File: trl/skills/trl-training/SKILL.md [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+
+**Basic usage:**
+
+```bash
+trl grpo \
+  --model_name_or_path Qwen/Qwen2.5-0.5B \
+  --dataset_name trl-lib/gsm8k \
+  --judge_model gpt-4 \
```
Comment (qgallouedec) [inline trl/skills/trl-training/SKILL.md]: same here, no judge model in grpo
